client/components/edit/StateModal.tsx

```diff
@@ -0,0 +1,188 @@
+import FormControlLabel from "@mui/material/FormControlLabel";
+import FormLabel from "@mui/material/FormLabel";
+import Grid from "@mui/material/Grid";
+import Radio from "@mui/material/Radio";
+import RadioGroup from "@mui/material/RadioGroup";
+import IconButton from "@mui/material/IconButton";
+import CloseIcon from "@mui/icons-material/Close";
+import { useEditorContext } from "../../context/EditorContext";
+import { getObjectRecursiveChildPidsUrl, getObjectStateUrl } from "../../util/routes";
+import { useFetchContext } from "../../context/FetchContext";
+import ObjectLoader from "./ObjectLoader";
+
+const StateModal = (): React.ReactElement => {
+    const {
+        state: { isStateModalOpen, objectDetailsStorage, stateModalActivePid },
+        action: { removeFromObjectDetailsStorage, setSnackbarState, toggleStateModal },
+    } = useEditorContext();
+    const {
+        action: { fetchJSON, fetchText },
+    } = useFetchContext();
+    const [statusMessage, setStatusMessage] = useState<string>("");
+    const [includeChildren, setIncludeChildren] = useState<boolean>(false);
+    const [selectedValue, setSelectedValue] = useState<string>("Inactive");
+    const [childPidResponse, setChildPidResponse] = useState({ loading: true });
+    const loaded = Object.prototype.hasOwnProperty.call(objectDetailsStorage, stateModalActivePid);
+    const details = loaded ? objectDetailsStorage[stateModalActivePid] : {};
+    const childPageSize = 1000;
+    useEffect(() => {
+        async function loadChildren() {
+            setChildPidResponse({ loading: true });
+            setIncludeChildren(false);
+            const url = getObjectRecursiveChildPidsUrl(details.pid, 0, childPageSize);
+            const response = await fetchJSON(url);
+            setChildPidResponse(response);
+        }
+        setSelectedValue(details.state ?? "Inactive");
+        if (details.pid ?? false) {
+            loadChildren();
+        }
+    }, [details]);
+
+    const handleRadioChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+        setSelectedValue(event.target.value);
+    };
+
+    const handleChildCheckboxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+        setIncludeChildren(event.target.checked);
+    };
+
+    const showSnackbarMessage = (message: string, severity: string) => {
+        setSnackbarState({
+            open: true,
+            message,
+            severity,
+        });
+    };
+
+    const updateStatus = async (pid: string): Promise<string> => {
+        setStatusMessage(`Saving status for ${pid}...`);
+        const target = getObjectStateUrl(pid);
+        const result = await fetchText(target, { method: "PUT", body: selectedValue });
+        if (result === "ok") {
+            // Clear and reload the cached object, since it has now changed!
+            removeFromObjectDetailsStorage(pid);
+        }
+        return result;
+    };
+
+    const saveChildPage = async (response): Promise<boolean> => {
```

Comment:
You can repurpose this to a Promise.all. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all

Reply:
I can change it to Promise.all with these diffs:

```
     const saveChildPage = async (response): Promise<boolean> => {
-        for (let i = 0; i < response.docs.length; i++) {
-            const result = await updateStatus(response.docs[i].id);
+        const results = await Promise.all(response.docs.map((doc) => updateStatus(doc.id)));
+        let success = true;
+        results.forEach((result) => {
             if (result !== "ok") {
                 showSnackbarMessage(`Status failed to save; "${result}"`, "error");
                 toggleStateModal();
                 setStatusMessage("");
-                return false;
+                success = false;
             }
-        }
-        return true;
+        });
+        return success;
     };
```

...however, I'm not sure if this is desirable. In the current logic, the work is done serially, which reduces the load on the server and allows clear progress messages. With Promise.all, the work is done in parallel, which potentially puts a heavier load on the server and makes it harder to create coherent messaging.

Reply:
Were you thinking of an application of Promise.all that retains the existing functionality but makes the code more elegant? Or are these changed characteristics inevitable if we do this refactoring?

Reply:
You don't necessarily need to do a forEach after. Promise.all will fail immediately when one of the promises fail. You can throw a particular error at the specific promise that is messing up. https://stackoverflow.com/questions/53981230/using-try-catch-with-promise-all.

Reply:
The forEach is not to detect failed promises, but rather to validate the API responses. The Promise.all runs a whole bunch of API calls and then returns an array of API responses. It's possible for the API call to fail without causing the promise to fail. I suppose I could put a check inside the updateStatus call to make the promise fail, instead of checking after the fact... but I still think we end up with the same "firehose" effect. If I get 1,000 IDs back, I'd rather work on those serially than try to make my browser simultaneously initiate 1,000 API calls. (And I realize that the browser will probably throttle that in some way... but I feel more comfortable controlling the flow explicitly in this case).

Reply:
The old logic iterates through the list and stops as soon as something goes wrong -- so it displays a snackbar message for the first failure. If we do things in parallel with Promise.all, we have less control over the flow of things. Array.some is a useful tool, but using it won't change the fact that the promises will all have been resolved before we can check the results. I think at this point I'm more inclined to leave this logic as it is, so we can stop the process if something goes wrong instead of potentially multiplying problems by doing too much work with the system in a bad state. I imagine we could potentially get a similar effect by throwing an exception inside the Promise.all callback, but that seems a bit less controlled and predictable than the current logic.